src/lib/tuning/index/base_index_evaluator.cpp

```diff
@@ -0,0 +1,222 @@
+      index_choice.memory_cost = static_cast<float>(_existing_memory_cost(index_choice));
+    } else {
+      index_choice.type = _propose_index_type(index_choice);
+      index_choice.memory_cost = static_cast<float>(_predict_memory_cost(index_choice));
+    }
+    index_choice.saved_work = _calculate_saved_work(index_choice);
+
+    // Transfer results to choices vector
+    choices.push_back(std::make_shared<IndexChoice>(index_choice));
+  }
+}
+
+void BaseIndexEvaluator::_setup() {}
+
+void BaseIndexEvaluator::_process_access_record(const BaseIndexEvaluator::AccessRecord&) {}
+
+uintptr_t BaseIndexEvaluator::_existing_memory_cost(const IndexChoice& index_choice) const {
+  const auto table = StorageManager::get().get_table(index_choice.column_ref.table_name);
+  uintptr_t memory_cost = 0u;
+  for (ChunkID chunk_id = ChunkID{0}; chunk_id < table->chunk_count(); ++chunk_id) {
+    const auto chunk = table->get_chunk(chunk_id);
+    const auto index = chunk->get_index(index_choice.type, index_choice.column_ref.column_ids);
+    if (index) {
+      memory_cost += index->memory_consumption();
+    }
+  }
+  return memory_cost;
+}
+
+void BaseIndexEvaluator::_inspect_query_cache() {
+  _access_records.clear();
+
+  /*
+   * ToDo(anybody): The cache interface could be improved by introducing values() method in
+   * AbstractCache interface and implement in all subclasses
+   *  const auto& query_plan_cache = SQLQueryOperator::get_query_plan_cache().cache();
+   * -> so this implementation could be independent of the actual cache implementation
+   */
+  const auto& lqp_cache = SQLQueryCache<std::shared_ptr<AbstractLQPNode>>::get();
+  // We cannot use dynamic_pointer_cast here because SQLQueryCache.cache() returns a reference, not a pointer
+  auto gdfs_cache_ptr =
+      dynamic_cast<const GDFSCache<std::string, std::shared_ptr<AbstractLQPNode>>*>(&(lqp_cache.cache()));
+  if (!gdfs_cache_ptr) {
+    LOG_WARN("BaseIndexEvaluator can only analyze GDFSCache instances! Evaluations may be useless...");
+    return;
+  }
+
+  const auto& fibonacci_heap = gdfs_cache_ptr->queue();
+
+  LOG_DEBUG("Query plan cache (size: " << fibonacci_heap.size() << "):");
+  if (fibonacci_heap.size() == 0) {
+    LOG_WARN("There are no logical query plans in the cache. Make sure that logical query plans get cached!");
+  }
+
+  auto cache_iterator = fibonacci_heap.ordered_begin();
+  auto cache_end = fibonacci_heap.ordered_end();
+
+  for (; cache_iterator != cache_end; ++cache_iterator) {
+    const auto& entry = *cache_iterator;
+    LOG_DEBUG("  -> Query '" << entry.key << "' frequency: " << entry.frequency << " priority: " << entry.priority);
+    _inspect_lqp_operator(entry.value, entry.frequency);
+  }
+}
+
+void BaseIndexEvaluator::_inspect_lqp_operator(const std::shared_ptr<const AbstractLQPNode>& op,
+                                               size_t query_frequency) {
+  std::list<const std::shared_ptr<const AbstractLQPNode>> queue;
```

Comment:
What‘s wrong with recursion?

Reply:
Nothing in particular here, we just decided to not do it in a recursive way. What's wrong with this way?

Reply:
It just looked like the prime example for recursion. Nothing wrong with it. ;)